### PR TITLE
Implement the new L1 attribute 'framework'

### PR DESCRIPTION
**Pull request by psss**
* Add specification for the new 'framework' attribute.
 * Implement the new atribute in the Test class.
 * Handle the old execution methods as framework default.
 * Adjust both internal and detach executors.
 * Include framework in run.yaml and adjust run.sh.
 * Include test coverage for both old and new methods.
 * Update execute step specification, add stories
 * Implement a new warn() method for user warnings.

### CODE REVIEW

**Comment by lukaszachy** [comment]
So we will break current plans (listed executors will not be available any more).
I can think only about keeping (for defined time being and leaving 'deprecated' message) current executors which will "ignore" this attribute - e.g. shell.tmt will ignore that tests are of type:beakerlib.

Personally I'd prefer to break it for current users, but since you need approval to modify plans doing this old executor hack might be worth it.

**Comment by pvalena** [comment]
> Personally I'd prefer to break it for current users, but since you need approval to modify plans doing this old executor hack might be worth it.

For backwards compatibility we could simply make L2 metadata entry change the _default_ for that run,

**Comment by lukaszachy** [comment]
> _default_

**Comment by lukaszachy** [comment]
Eh, sorry for closing - wrong button

**Comment by psss** [comment]
Thinking a bit about the attribute name. Shall we perhaps choose a different name than `type` as the term `test type` is often used to describe the objective of testing? ISTQB says:

> A test type is focused on a particular test objective, which could be the testing of a function to be performed by the component or system; a non-functional quality characteristic, such as reliability or usability; the structure or architecture of the component or system; or related to changes, i.e. confirming that defects have been fixed (confirmation testing, or re-testing) and looking for unintended changes (regression testing).

So for this attribute the relevant values could be:

* functional
* non-functional
* structural
* confirmation
* regression

Is this something we will possibly want to track in the future? Would `framework` be a better name? Any other ideas?

**Comment by thrix** [comment]
`framework` sound better to me

**Comment by psss** [comment]
Based on the feedback from stakeholders, let's use `framework` as the name for the new attribute.

**Comment by psss** [comment]
For reference, adding a couple of definitions of a testing framework:

> A test automation framework is an integrated system that sets the rules of automation of a specific product. This system integrates the function libraries, test data sources, object details and various reusable modules. These components act as small building blocks which need to be assembled to represent a business process. The framework provides the basis of test automation and simplifies the automation effort. 

The Testing framework is responsible for:

 * defining the format in which to express expectations
 * creating a mechanism to hook into or drive the application under test
 * executing the tests
 * reporting results

Defining boundaries between automation framework and a testing tool

> Tools are specifically designed to target some particular test environment, such as Windows and web automation tools, etc. Tools serve as a driving agent for an automation process. However, an automation framework is not a tool to perform a specific task, but rather infrastructure that provides the solution where different tools can do their job in a unified manner. This provides a common platform for the automation engineer.

There are various types of frameworks. They are categorized on the basis of the automation component they leverage. These are:

 * Data-driven testing
 * Modularity-driven testing
 * Keyword-driven testing
 * Hybrid testing
 * Model-based testing
 * Code-driven testing
 * Behavior driven development

A couple of related links:

 * https://en.wikipedia.org/wiki/Test_automation
 * https://smartbear.com/learn/automated-testing/test-automation-frameworks/

File: spec/tests/framework.fmf
Comment (thrix) [inline spec/tests/framework.fmf:21]: should be `framework:`

(also the next one)

Comment (psss) [reply spec/tests/framework.fmf]: Ah, thanks. Fixed.

**Comment by psss** [comment]
@thrix, @lukaszachy, @pvalena, @FrNecas, @hegerj, the implementation should be done. Ready for your review. As this is rather a bit larger change which substantially affects the core functionality of test execution I would welcome if you could give it a bit of testing as well. Thanks.

**Comment by psss** [comment]
Regarding the attribute name: The more I'm thinking about it and the more times I was writing it in the patch, the more I realized I don't like much the term `framework`. As provided a couple of links and definitions above, the term "test automation framework" is used in much more broader context. I did another short research and all seem to say something similar:

https://www.softwaretestinghelp.com/test-automation-frameworks-selenium-tutorial-20/
> A framework is a constructive blend of various guidelines, coding standards, concepts, processes, practices, project hierarchies, modularity, reporting mechanism, test data injections etc. to pillar automation testing. Thus, the user can follow these guidelines while automating application to take advantages of various productive results.

https://www.guru99.com/test-automation-framework.html
> A set of guidelines like coding standards , test-data handling , object repository treatment etc... which when followed during automation scripting produce beneficial outcomes like increase code re-usage , higher portability , reduced script maintenance cost etc.

https://www.softwaretestingmaterial.com/types-test-automation-frameworks/
> A framework defines a set of rules or best practices that we can follow in a systematic way to achieve the desired results.

The best alternative term which comes to my mind is a "test automation tool", so `tool` could be the name of the attribute. Although it's also quite generic, I'd say it better fits to say `shell` is a tool used for testing rather then `shell` is a testing framework. Similarly for `beakerlib` I'd say "tool" fits better than "framework". And for both definitely "test tool" sounds better than the original "test type".

Ideas? Comments? Other term proposals?

**Comment by sopos** [comment]
I don't think `beakerlib` or `shell` are really tools. It is definitely better than just a `type` but I do not like it neither. It would be good to have an opinion of some native English speaker.

**Comment by ep69** [comment]
I don't like the options so far. What about `format`?

**Comment by pvalena** [comment]
> The best alternative term which comes to my mind is a "test automation tool", so `tool` could be the name of the attribute. Although it's also quite generic, I'd say it better fits to say `shell` is a tool used for testing rather then `shell` is a testing framework. Similarly for `beakerlib` I'd say "tool" fits better than "framework". And for both definitely "test tool" sounds better than the original "test type".
> 
> Ideas? Comments? Other term proposals?

I actually liked "test type" (and yes "framework" _keyword_ is taken, so it could be confusing for some), as it expresses what it actually is, no? Does the "type/framework/tool" actually imply any enforced rules / checks / or behaviour changes? If not, we could call it simply "category". `Tool` is IMO even more generic, and i collides with the name of this project, so `-1` for me.

File: stories/cli/steps.fmf
Comment (optak) [inline stories/cli/steps.fmf:88]: I think that it would be good to preserve "shell" example.

**Comment by sopos** [comment]
> I don't like the options so far. What about `format`?

This is actually the only option I could come up with so far as well.

**Comment by pvalena** [comment]
To iterate: Framework is actually a very fitting name, because we will execute a test via some -external from tmt's POV- `framework`; that's the idea IMO, to encapsulate runtime of other, already developed frameworks, parse the output and aggregate it.

**Comment by hhorak** [comment]
Thinking about more examples than beakerlib and shell, like pytest, some refer to themselves like frameworks as well. Also this page makes me think `framework` is not actually that bad: https://en.wikipedia.org/wiki/List_of_unit_testing_frameworks
Another synonym that would make sense to me would be `library`.

**Review by optak — APPROVED** [review]
LGTM

File: stories/cli/steps.fmf
Comment (psss) [reply stories/cli/steps.fmf:88]: Why? We want to deprecate the old execute methods. So I'd say the documentation should not recommend them in any way.

**Comment by sopos** [comment]
> Thinking about more examples than beakerlib and shell, like pytest, some refer to themselves like frameworks as well. Also this page makes me think `framework` is not actually that bad: https://en.wikipedia.org/wiki/List_of_unit_testing_frameworks
> Another synonym that would make sense to me would be `library`.

A `framework` is definitely not that bad, definitely better that a `tool`. But a `library` would be misleading ever more.
Currently my favourites are `framework` and `format`.

**Comment by odubaj** [comment]
What about "instrument" ? From the meaning: a tool or implement, especially one for precision work

**Comment by odubaj** [comment]
we are using instruments while comleting a job. So we can use an instrument(shell/beakerlib) to complete a job(test-run). Does it make sense ?

**Comment by sopos** [comment]
Isn't an `instrument` a synonym of a `tool`?

**Comment by odubaj** [comment]
Yes, but we are trying to find the best possible name for the attribute, right? For me "instrument" sounds better.

**Comment by junaruga** [comment]
I just love seeing people discussing the attribute name here. That's why tmt's user interface is sophisticated.

File: stories/cli/steps.fmf
Comment (optak) [reply stories/cli/steps.fmf:88]: Ok, since shell is default framework, and it's nicely documented, similar example (of framework) is probably not needed. Here I mixed "framework" and "how" in my head.

**Comment by lukaszachy** [comment]
I vote for `framework` since it follows, imo, Principle of least astonishment - values of this attribute are expected to be test frameworks + shell.

**Comment by i386x** [comment]
Maybe `way`, but it is also so generic.

**Comment by sopos** [comment]
what about following approach:
* propose a name as s separate comment (the name only in the text)
* use thumbs up / down to any proposal you like / dislike
* accept proposals to some date
* gather the votes by another date

**Comment by sopos** [comment]
type

**Comment by sopos** [comment]
framework

**Comment by sopos** [comment]
format

**Comment by sopos** [comment]
tool

**Comment by sopos** [comment]
instrument

**Comment by sopos** [comment]
way

**Comment by sopos** [comment]
library

**Comment by psss** [comment]
method

**Comment by psss** [comment]
The format sounds really good to me. Thanks for the idea, @ep69. Let's see how it would sound:

 * If the test comes in the `format` of a shell script we just execute it and check for the exit code.
 * If the test is written in the beakerlib `format` we install additional dependency and check the TestResults file.

Or, we could get rid of two letters and use `form` which can be defined as "the way in which a thing exists, acts, or manifests itself" :-)

**Comment by psss** [comment]
form

**Comment by pvalena** [comment]
> method

_Comment: this could collide with some keywords; and also is synonymous of `how` we have already reserved for the `execute` step itself._

**Comment by sopos** [comment]
> form

this asks for a typo `from` and one would be surprised

**Comment by sopos** [comment]
> > method
> 
> _Comment: this could collide with some keywords; and also is synonymous of `how` we have already reserved for the `execute` step itself._

`method` is not bad to me and it correspond with `provision --how` so I would not refuse it just because of the synonymousness with `how`

**Comment by pvalena** [comment]
> `method` is not bad to me and it correspond with `provision --how` so I would not refuse it just because of the synonymousness with `how`

That's what I mean. We have already `execute --how` as well (every step has "how").

**Comment by sopos** [comment]
> > `method` is not bad to me and it correspond with `provision --how` so I would not refuse it just because of the synonymousness with `how`
> 
> That's what I mean. We have already `execute --how` as well (every step has "how").

I see your point (I wanted to say `execute` instead of `provision`) but you're right that we might want to have something in L1 to express a need for specific `provision` way or whatever in which case it might collide.

**Comment by pvalena** [comment]
> I see your point (I wanted to say `execute` instead of `provision`) but you're right that we might want to have something in L1 to express a need for specific `provision` way or whatever in which case it might collide.

Sorry, I don't follow. Could you elaborate?

AFAIK L1 should never have any metadata for provisioning - because provisioning is simply ""running the VM"". When it runs, it's up to a `prepare` step. So all provisioning metadata need to be in L2.

**Comment by sopos** [comment]
> Sorry, I don't follow. Could you elaborate?
> 
> AFAIK L1 should never have any metadata for provisioning - because provisioning is simply ""running the VM"". When it runs, it's up to a `prepare` step. So all provisioning metadata need to be in L2.
> 
> Also it might be worth noting there can be different implementations / variations for provisioning, like `virtual.libvirt.vagrant` (I hope I remembered the order correctly).

I hope there will not be any significant need to for a specific provision method per test. However, we thought it will not be needed even for the `execute` method which appears to be really a must for some use cases. So who knows...
There was probably some misunderstanding between @sopos and @pvalena but I think it was not about anything important so let's not deep dive into it any more while it is in a theoretical level.

**Comment by psss** [comment]
> I vote for `framework` since it follows, imo, Principle of least astonishment - values of this attribute are expected to be test frameworks + shell.

I'd say the main problem with `framework` is that the term doesn't match not only for `shell` but also does not fit for `beakerlib` which is rather a library. Looking at the next hot candidate, [behave](https://behave.readthedocs.io/en/latest/)'s homepage does not mention framework at all.

On the other hand, I see that on other places behave is described as a "behavior-driven test framework" and [dogtail](https://gitlab.com/dogtail/dogtail), which could be another possible candidate, defines itself as a "test tool and automation framework".

So after the discussion so far, my current top candidates are "test format", "test tool" and "test framework". Let's see how they would look like in action:

    format: shell
    format: beakerlib
    format: dogtail
    format: behave

    tool: shell
    tool: beakerlib
    tool: dogtail
    tool: behave

    framework: shell
    framework: beakerlib
    framework: dogtail
    framework: behave

**Comment by sopos** [comment]
> ```
> format: shell
> format: beakerlib
> format: dogtail
> format: behave
> 
> tool: shell
> tool: beakerlib
> tool: dogtail
> tool: behave
> 
> framework: shell
> framework: beakerlib
> framework: dogtail
> framework: behave
> ```

`tool` is the most "hated" out of these three so I would cut it off. Personally I would pick `format` even if it is not winning by votes but it is you who has the final word, @psss.

**Comment by pkis** [comment]
If it is not too late, consider also: `interpreter`

**Comment by i386x** [comment]
What about `mode`?
```
summary: Test execution mode defining how tests should be executed

story:
    As a tester I want to include tests using different test
    execution mode in a single plan.
```

**Comment by sopos** [comment]
`interpreter`

**Comment by sopos** [comment]
`mode`

**Comment by psss** [comment]
Reminder: Last 24 hours to share your feedback regarding the new attribute name. Thanks.

**Comment by pvalena** [comment]
> > I vote for `framework` since it follows, imo, Principle of least astonishment - values of this attribute are expected to be test frameworks + shell.
> 

IMO for `shell` it fits as well, because we do not necessarily care about what the insides of the tests are, but rather: how the test is executed and how the result is interpreted (IOW based on `exit code` for `shell`).

**Comment by pvalena** [comment]
I did some testing (in VM, that's why some tests fail on timeout). I've also run the test execution with this PR (`tmt-0.21-1.20200923150010200303.pr302.22.g84b34dd.fc34.src.rpm`).

```
$ tmt run -avv plan --name=/try/container/ --name=/try/virtual
[  . . . ]
total: 65 tests passed, 6 tests failed and 10 errors
```
[Full log](https://git.io/JUDrD).

With some [minor modifications](https://git.io/JUDrz).

The master yields the same number of errors. => LGTM.

**Comment by psss** [comment]
Thanks for the testing, Pavel.
@pvalena, @lukaszachy, @thrix, any feedback to the code itself?

File: spec/steps/execute.fmf
Comment (thrix) [inline spec/steps/execute.fmf:75]: typo: `life` -> `live` :)

File: stories/cli/steps.fmf
Comment (thrix) [inline stories/cli/steps.fmf:117]: I would mention here that this makes it possible to insert directly `bash` in the test to do interactive debugging, might not be abvious.

File: tmt/steps/execute/__init__.py
Comment (thrix) [inline tmt/steps/execute/__init__.py:67]: And will be removed by March 2021? :)

Comment (thrix) [reply tmt/steps/execute/__init__.py:67]: seems quite an important piece of info

Comment (thrix) [reply tmt/steps/execute/__init__.py:67]: what about adding here a check fo March 2021 and just error in the case the date has passed? This way users will be nicely informed with an error that they are using a deprecated functionality

File: spec/steps/execute.fmf
Comment (psss) [reply spec/steps/execute.fmf:75]: Nice! :) Fixed.

File: stories/cli/steps.fmf
Comment (psss) [reply stories/cli/steps.fmf:117]: Good point. Added an explicit mention.

File: tmt/steps/execute/__init__.py
Comment (psss) [reply tmt/steps/execute/__init__.py:67]: Well this is showing a warning message for each obsolete config already now:

```
> tmt run plan --name plans/smoke
/var/tmp/tmt/run-478

/plans/smoke
    warning: The 'shell.tmt' execute method has been deprecated.
    warning: Use 'how: tmt' in the execute step instead (L2).
    warning: Set 'framework: shell' in test metadata (L1).
    warning: Support for old methods will be dropped in March 2021.
    discover
        how: shell
        directory: /home/psss/git/tmt
        summary: 1 test selected
    provision
        how: local
        distro: Fedora 32 (Workstation Edition)
        summary: 1 guest provisioned
    prepare
        summary: 0 preparations applied
    execute
        how: tmt
        summary: 1 test executed
    report
        how: display
        summary: 1 test passed
    finish
        summary: 0 tasks completed

total: 1 test passed
```

In March I would like to completely remove the functionality and fail with an error.

Comment (thrix) [reply tmt/steps/execute/__init__.py:67]: agreed, I was just proposing to add the check now .. .but it is your choice :+1:

**Comment by thrix** [comment]
I guess the update is jut about to land, I do not see it yet

File: tmt/steps/execute/__init__.py
Comment (psss) [reply tmt/steps/execute/__init__.py:67]: Also, before completely removing the feature I would rather check the current situation at that time and announce this backward incompatible change so that users are not surprised.

**Review by thrix — APPROVED** [review]
mvadkert is happy now

**Comment by psss** [comment]
The decision regarding the name was not easy. For quite some time I was convinced that `format` was the best candidate, but let's go forward with the `framework`. In some respects it might be still confusing but despite this it seems to be the best term we can make up which is closely associated with automated testing tools. The last drop which contributed to the final decision was search for [test format](https://www.google.com/search?q="test+format") versus [test framework](https://www.google.com/search?q=%22test+framework%22) which suggests that test format is rather connected with test questions.

Thanks much to all for review, testing and providing input regarding the new attribute name. I've run the whole test coverage against all provision methods. All 164 tests passed, we're good to go. Merging.

**Pull request merged**
